# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_target.rs
# Commit: c20cf2fa829a
# Full Hash: c20cf2fa829a19cb6b60909d05abcda40e77dd70
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2022-04-20 09:42:47
# Regressor Bug: 1764005
# File Overlap Count: 1
# Description:
#   Bug 1764005 - Add backdrop-filter primitive support r=gfx-reviewers,lsalzman,nical
#   
#   Add support for backdrop-filter primitive in WR. Each backdrop
#   filter primitive establishes a render task sub-graph. The primitive
#   instance registers itself as a resolve source to receive the
# ==============================================================================

diff -r d5b01209cf78 -r c20cf2fa829a gfx/wr/webrender/src/render_target.rs
--- a/gfx/wr/webrender/src/render_target.rs	Wed Apr 20 01:08:52 2022 +0000
+++ b/gfx/wr/webrender/src/render_target.rs	Wed Apr 20 01:09:09 2022 +0000
@@ -222,7 +222,7 @@
     // we can set a scissor rect and only clear to the
     // used portion of the target as an optimization.
     pub used_rect: DeviceIntRect,
-    pub resolve_op: Option<ResolveOp>,
+    pub resolve_ops: Vec<ResolveOp>,
     pub clear_color: Option<ColorF>,
 }
 
@@ -244,7 +244,7 @@
             screen_size,
             texture_id,
             used_rect,
-            resolve_op: None,
+            resolve_ops: Vec::new(),
             clear_color: Some(ColorF::TRANSPARENT),
         }
     }
@@ -277,8 +277,8 @@
                         Some(target_rect)
                     };
 
-                    if pic_task.is_resolve_target {
-                        self.clear_color = None;
+                    if !pic_task.can_use_shared_surface {
+                        self.clear_color = pic_task.clear_color;
                     }
 
                     // TODO(gw): The type names of AlphaBatchBuilder and BatchBuilder
@@ -371,9 +371,9 @@
                 );
             }
             RenderTaskKind::Picture(ref pic_task) => {
-                // Only one resolve per target is supported for now
-                debug_assert!(self.resolve_op.is_none());
-                self.resolve_op = pic_task.resolve_op;
+                if let Some(ref resolve_op) = pic_task.resolve_op {
+                    self.resolve_ops.push(resolve_op.clone());
+                }
                 self.alpha_tasks.push(task_id);
             }
             RenderTaskKind::SvgFilter(ref task_info) => {
@@ -396,6 +396,7 @@
             RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
+            RenderTaskKind::TileComposite(..) |
             RenderTaskKind::LineDecoration(..) => {
                 panic!("Should not be added to color target!");
             }
@@ -492,6 +493,7 @@
             RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
+            RenderTaskKind::TileComposite(..) |
             RenderTaskKind::SvgFilter(..) => {
                 panic!("BUG: should not be added to alpha target!");
             }
@@ -569,22 +571,33 @@
 
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
-#[derive(Debug, PartialEq, Copy, Clone)]
+#[derive(Debug, PartialEq, Clone)]
 pub struct ResolveOp {
-    pub src_task_id: RenderTaskId,
+    pub src_task_ids: Vec<RenderTaskId>,
     pub dest_origin: DevicePoint,
     pub dest_task_id: RenderTaskId,
 }
 
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
+pub enum PictureCacheTargetKind {
+    Draw {
+        alpha_batch_container: AlphaBatchContainer,
+    },
+    Blit {
+        task_id: RenderTaskId,
+        sub_rect_offset: DeviceIntVector2D,
+    },
+}
+
+#[cfg_attr(feature = "capture", derive(Serialize))]
+#[cfg_attr(feature = "replay", derive(Deserialize))]
 pub struct PictureCacheTarget {
     pub surface: ResolvedSurfaceTexture,
-    pub alpha_batch_container: AlphaBatchContainer,
+    pub kind: PictureCacheTargetKind,
     pub clear_color: Option<ColorF>,
     pub dirty_rect: DeviceIntRect,
     pub valid_rect: DeviceIntRect,
-    pub resolve_op: Option<ResolveOp>,
 }
 
 #[cfg_attr(feature = "capture", derive(Serialize))]
@@ -704,6 +717,7 @@
             RenderTaskKind::CacheMask(..) |
             RenderTaskKind::Readback(..) |
             RenderTaskKind::Scaling(..) |
+            RenderTaskKind::TileComposite(..) |
             RenderTaskKind::SvgFilter(..) => {
                 panic!("BUG: unexpected task kind for texture cache target");
             }